sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs

```diff
@@ -0,0 +1,129 @@
+        private const int DefaultCopyBufferSize = 81920;
+
+        private readonly TimeSpan _networkTimeout;
+
+        public ResponseBodyPolicy(TimeSpan networkTimeout)
+        {
+            _networkTimeout = networkTimeout;
+        }
+
+        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            await ProcessAsync(message, pipeline, true).ConfigureAwait(false);
+        }
+
+        public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessAsync(message, pipeline, false).EnsureCompleted();
+        }
+
+        private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
+        {
+            CancellationToken oldToken = message.CancellationToken;
+            using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(oldToken);
+
+            cts.CancelAfter(_networkTimeout);
+
+            try
+            {
+                message.CancellationToken = cts.Token;
+                if (async)
+                {
+                    await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+                }
+                else
+                {
+                    ProcessNext(message, pipeline);
+                }
+            }
+            finally
+            {
+                message.CancellationToken = oldToken;
+                cts.CancelAfter(Timeout.Infinite);
+            }
+
+            Stream? responseContentStream = message.Response.ContentStream;
+            if (responseContentStream == null || responseContentStream.CanSeek)
+            {
+                return;
+            }
+
+            if (message.BufferResponse)
+            {
+                var bufferedStream = new MemoryStream();
+                if (async)
+                {
+                    await CopyToAsync(responseContentStream, bufferedStream, cts).ConfigureAwait(false);
+                }
+                else
+                {
+                    CopyTo(responseContentStream, bufferedStream, message.CancellationToken);
+                }
+
+                responseContentStream.Dispose();
+                bufferedStream.Position = 0;
+                message.Response.ContentStream = bufferedStream;
+            }
+            else if (_networkTimeout != Timeout.InfiniteTimeSpan)
+            {
+                message.Response.ContentStream = new ReadTimeoutStream(responseContentStream, _networkTimeout);
+            }
+        }
+
+        private async Task CopyToAsync(Stream source, Stream destination, CancellationTokenSource cancellationTokenSource)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(DefaultCopyBufferSize);
+            try
+            {
+                while (true)
+                {
+                    cancellationTokenSource.CancelAfter(_networkTimeout);
```

Comment:
Shouldn't we set it before loop?

Reply:
It's a per-read timeout so it's set per iteration.

Reply:
Should there be similar checks in `CopyTo`?

Reply:
Stream.Read/Write do not take cancellation tokens anyway so it's impossible to cancel them while they are running.